### PR TITLE
Add a minimal "notes" feature to Books / Reading Log

### LINKED ISSUE #4294: Add ability to add personal Notes to Books
### Describe the problem that you'd like solved

Right now when people recommend a book to me, I want to note somewhere who recommended the book. I'd also like to be able to jot down notes about specific editions.

### Proposal & Constraints

Decided to create a new table rather than to piggy-back on Reading Log, which, could have performance effects and other undesirable side effects.

Not supposed to be a large feature, as simple as possible to get something functional / "possible"

### PR DESCRIPTION
**Pull request by mekarpeles**
Closes #4294

Extends @jamesachamp's work on #4135 to allow patrons to add "notes" to individual books in their Reading Log. 

Notes form piggy backs on the modal for the https://github.com/Open-Book-Genome-Project/TheBestBookOn.com patron metadata form for book characterisics.

### Technical

Note: A text column has already been added to ol-db1 / prod database (as has the corresponding sql to openlibrary's `schema.sql`).

We've also updated the core/bookshelves.py model to pass a `note` through to the DB. There isn't yet an endpoint for update a Reading Log entry to only add a note.

### Discussion & Considerations

Right now, you can only add a note if you're in the beta program and if you've added a book to your reading log. Unclear (in current proposed implementation) why notes tied to Reading Log and (if it stays that way) what should happen to notes if someone removes book from reading log (probably a compelling reason to create a new table like `patron_notes`, though this may complicate things a sit will require additional joins in order to show notes on the patron's Reading Log page. Not the end of the world. In any even relatively easy to change and designed for beta).

### Testing

Nothing to test yet 

### Screenshot

dev openlibrary org_books_OL24755024M_The_untethered_soul

dev openlibrary org_books_OL24755024M_The_untethered_soul (1)

### Stakeholders

@jamesachamp, @seabelis

### CODE REVIEW

File: openlibrary/core/booknotes.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
+from . import db
+
+
+class Booknotes(object):
+
+    @classmethod
+    def total_booknotes(cls):
+        oldb = db.get_db()
+        query = "SELECT count(*) from booknotes"
+        return oldb.query(query, vars=data)['count']
```
Comment (mekarpeles) [inline openlibrary/core/booknotes.py:10]:
```suggestion
        return oldb.query(query)['count']
```

File: openlibrary/core/booknotes.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    @classmethod
+    def most_notable_books(cls, limit=10, since=False):
+        """Across all patrons"""
+        oldb = db.get_db()
+        query = "select work_id, count(*) as cnt from booknotes"
+        if since:
+            query += " AND created >= $since"
+        query += ' group by work_id order by cnt desc limit $limit'
+        return list(oldb.query(query, vars={
+            'shelf_id': shelf_id,
```
Comment (mekarpeles) [inline openlibrary/core/booknotes.py:41]:
```suggestion
```

File: openlibrary/plugins/openlibrary/api.py [hunk 3]
```diff
@@ -121,6 +122,54 @@ def response(msg, status="success"):
...
+        :param str work_id: e.g. OL123W
+        :rtype: json
+        :return: the note
+        """
+        user = accounts.get_current_user()
+        i = web.input(notes=None, edition_id=None, redir=None)
+        edition_id = int(extract_numeric_id_from_olid(i.edition_id)) if i.edition_id else None
+
+        if not user:
+            raise web.seeother('/account/login?redirect=%s' % key)
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py:146]:
```suggestion
            raise web.seeother('/account/login?redirect=/works/%s' % work_id)
```

File: openlibrary/plugins/openlibrary/api.py [hunk 4]
```diff
@@ -121,6 +122,54 @@ def response(msg, status="success"):
...
+
+        Booknotes.add(
+            username=username,
+            work_id=work_id,
+            notes=i.notes,
+            edition_id=edition_id
+        )
+
+        if i.redir:
+            raise web.seeother(key)
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py:167]:
```suggestion
            raise web.seeother("/works/%s" % work_id)
```

**Review by mekarpeles — COMMENTED** [review]
CI issues, undefined vars

**Comment by mekarpeles** [comment]
For whomever assigns to this issue **please squash on merge** :) thank you!

**Comment by mekarpeles** [comment]
Merging to unblock @jamesachamp

File: openlibrary/core/booknotes.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    @classmethod
+    def remove(cls, username, work_id):
+        """Remove a patron's specific booknote by work_id.
+
+        Technical note: work_id is not an optional argument and
+        intentionally does not default to None (to reduce
+        accidents/risk), however if one passes None as a value to
+        work_id, this method will remove all booknotes for a patron
+        (useful for a patron who may decide to close their account.
```
Comment (jamesachamp) [inline openlibrary/core/booknotes.py:125]: Is this accurate?  `work_id` is a primary key field, so it is not nullable by default.

Casting `None` to an int (at line 134) will cause a TypeError, as well.

If any of this should be changed, I can work on it while I make booknotes edition-level.

**Pull request merged**